### ISSUE TITLE
Job terminated with signal 11 at PC=xxx error

### ISSUE THREAD

**Issue by Chris-Pedersen** [rainwoodman/MP-Gadget#171]
Myself and @andreufont have had some Lya sims crash with the error:

`
MP-Gadget:33262 terminated with signal 11 at PC=3a96089f7e SP=7ffc8bfc3ba8.
`

The Makefile and full output files are here:
[test_err.txt](https://github.com/rainwoodman/MP-Gadget/files/1531637/test_err.txt)
[test_out.txt](https://github.com/rainwoodman/MP-Gadget/files/1531639/test_out.txt)
[Optionsmk.txt](https://github.com/rainwoodman/MP-Gadget/files/1531655/Optionsmk.txt)

If I increase PartAllocFactor to 4 the sims work ok, but these sims were running fine a few weeks ago with it set to 2. Should changing PartAllocFactor be necessary for the params (L=10Mpc, Ngrid=256) we are using?

Thanks!

**Comment by rainwoodman** [comment]
The slotsmanager was merged a few weeks ago to allow dynamically increasing the number of particles.

Setting partallocfacter too small shouldn't cause the code to crash like this -- if it is too small during domain decomposition we are supposed to print an error and die.

The crash on a memmove is likely the one in slotsmanager.c. The last message says rank 37 is trying to increase the number of slots. 11 is Segmental fault. So somehow the memmove there is writing to bad addresses. Though I don't immediately see how -- the numbers all looked reasonable. It would be the best if we can get a stack trace -- what if you run with the code compiled with '-Og -g'?

Also, do you have a checkpoint / snapshot that you can restart from? Does the problem go away if you restart with alloc=2?

**Comment by sbird** [comment]
This may be fixed by commit 7d07b50e47 so try git-master

**Comment by Chris-Pedersen** [comment]
I tried with the latest commit and I get the same error - I'll try Yu's suggestions tomorrow.

I also noticed in the output it says it's allocating memory for BHs & stars, even right at the very start of the sim:

`[ 000005.08 ] Task 1: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.`

even though I compiled and ran with black holes off:

`BlackHoleOn                     0                    # Enable Blackhole`

Is this all ok?

**Comment by Chris-Pedersen** [comment]
I'm unable to restart from snapshots - I get:

[ 000000.01 ] Task 0: Last Snapshot number is 6.
...
[ 000000.00 ] Task 29: Last Snapshot number is 6.
[ 000000.01 ] Nhost = 2
[ 000000.01 ] Reserving 6774237866 bytes per rank for MAIN memory allocator.
[ 000004.26 ] Reserving 4206592 bytes per rank for TEMP memory allocator.
[ 000004.26 ] Aggregated IO is disabled.
[ 000004.85 ] Probing Header of snapshot file: output/snap_006
[ 000004.87 ] Failed to open snapshot at output/snap_006:Big File `output/snap_006' does not exist (No such file or directory) @(bigfile.c:172)

Am I saving snapshots in the wrong way to restart them? My output folder contains PIG_ and PART_ folders, but nothing with snap_.

I also tried recompiling with '-Og -g' and nothing changed - should I be running anything differently with these flags?

And finally, how can I provide a stack trace? There's a .btr file alongside the outputs, but that just contains the same information that's in the .err file.

**Comment by rainwoodman** [comment]
Looks like the code is trying restart with snap_xxx file. I think you've just stepped to another bug about snapshot writing. try to symlink PART_006 to snap_006

Getting a stack trace depends on the platform. Usually if the code is compiled with -g it should have add line numbers in the trace. But if it doesn't we can't do much... Does it produce a core dump file? we can use gdb to attach to the core file and get a stack trace. (my laptop is out of bat if this sounds too brief I can add more later).

**Comment by Chris-Pedersen** [comment]
I'll have a look at stack tracing tomorrow - is it possible/likely that this is an issue with the cluster we're using and not the code?

**Comment by rainwoodman** [comment]
You can ask the sysadmin on ways of obtaining a core dump file. Actually I think you can copy the core dump and the source code tree and the binary file to me, such that I can look at it.

**Comment by sbird** [comment]
I wonder if we should use libunwind (http://www.nongnu.org/libunwind/) to add custom code to generate a backtrace inside endrun (and a segfault handler): that might help with debugging.

**Comment by sbird** [comment]
PS I fixed the "unable to restart from snapshot" problem with b66712c0527

**Comment by sbird** [comment]
@Chris-Pedersen : It allocates some memory for black holes at simulation start in case some form, somehow. We could probably avoid doing this if All.BlackHolesOn = 0, but we didn't get around to it yet (since it is a relatively small amount of memory). You're welcome to implement the feature in a pull request if you like.

**Comment by sbird** [comment]
Several of the ranks are repeatedly reallocing to the same number of slots (see below). 
I don't think this should happen, unless:
a) it is increasing maxsize for some non-enabled slots (somehow? This does not seem possible)
b) there is some kind of memory error which is corrupting maxsize, resetting it to zero?

I have done some edits in the slotsfix branch (see https://github.com/rainwoodman/MP-Gadget/pull/174) which probably won't fix the problem, but should give us enough information to work out which it is.

eg, task 43:

[sbird4]~/Downloads% grep "Task 43: Allocated " test_out-1.txt   
[ 000005.08 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 000786.04 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 001431.72 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 001532.97 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 001939.27 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 002146.13 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 002413.44 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 002690.05 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 002932.12 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 003191.52 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 003414.95 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 003487.40 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 003784.57 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 004065.79 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 004281.60 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 004387.77 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 004500.33 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 004613.94 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 004729.69 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 004847.43 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 004966.08 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 005243.27 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 005400.67 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 005705.61 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 006046.25 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 006048.60 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 006079.89 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 006599.10 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 007021.06 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 007346.64 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 007444.89 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 007993.88 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 008453.94 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 008876.75 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 009269.04 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 009661.63 ] Task 43: Allocated 107.413 MB for 713031 sph, 27962 stars and 13981 BHs.
[ 009843.62 ] Task 43: Allocated 108.245 MB for 713031 sph, 49767 stars and 13981 BHs.
[ 009982.57 ] Task 43: Allocated 108.245 MB for 713031 sph, 49767 stars and 13981 BHs.
[ 010134.79 ] Task 43: Allocated 108.245 MB for 713031 sph, 49767 stars and 13981 BHs.
[ 010377.50 ] Task 43: Allocated 108.245 MB for 713031 sph, 49767 stars and 13981 BHs.
[ 010516.15 ] Task 43: Allocated 108.512 MB for 713031 sph, 56768 stars and 13981 BHs.
[ 010619.43 ] Task 43: Allocated 108.512 MB for 713031 sph, 56768 stars and 13981 BHs.
[ 010881.74 ] Task 43: Allocated 108.512 MB for 713031 sph, 56768 stars and 13981 BHs.
[ 010921.49 ] Task 43: Allocated 108.512 MB for 713031 sph, 56768 stars and 13981 BHs.
[ 011003.07 ] Task 43: Allocated 108.512 MB for 713031 sph, 56768 stars and 13981 BHs.
[ 011085.88 ] Task 43: Allocated 108.512 MB for 713031 sph, 56768 stars and 13981 BHs.
[ 011159.39 ] Task 43: Allocated 108.512 MB for 713031 sph, 56768 stars and 13981 BHs.
[ 011243.21 ] Task 43: Allocated 108.512 MB for 713031 sph, 56768 stars and 13981 BHs.
[ 011321.23 ] Task 43: Allocated 108.512 MB for 713031 sph, 56768 stars and 13981 BHs.
[ 011406.21 ] Task 43: Allocated 108.512 MB for 713031 sph, 56768 stars and 13981 BHs.
[ 011548.34 ] Task 43: Allocated 108.779 MB for 713031 sph, 63762 stars and 13981 BHs.
[ 011561.35 ] Task 43: Allocated 108.779 MB for 713031 sph, 63762 stars and 13981 BHs.
[ 011651.63 ] Task 43: Allocated 108.779 MB for 713031 sph, 63762 stars and 13981 BHs.
[ 011927.84 ] Task 43: Allocated 108.779 MB for 713031 sph, 63762 stars and 13981 BHs.
[ 012185.92 ] Task 43: Allocated 108.779 MB for 713031 sph, 63762 stars and 13981 BHs.
[ 012364.90 ] Task 43: Allocated 109.047 MB for 713031 sph, 70779 stars and 13981 BHs.
[ 012454.08 ] Task 43: Allocated 109.047 MB for 713031 sph, 70779 stars and 13981 BHs.
[ 012512.44 ] Task 43: Allocated 109.047 MB for 713031 sph, 70779 stars and 13981 BHs.
[ 012715.70 ] Task 43: Allocated 109.314 MB for 713031 sph, 77796 stars and 13981 BHs.
[ 012733.33 ] Task 43: Allocated 109.314 MB for 713031 sph, 77796 stars and 13981 BHs.
[ 012916.28 ] Task 43: Allocated 109.314 MB for 713031 sph, 77796 stars and 13981 BHs.
[ 013084.13 ] Task 43: Allocated 109.581 MB for 713031 sph, 84790 stars and 13981 BHs.
[ 013146.24 ] Task 43: Allocated 109.581 MB for 713031 sph, 84790 stars and 13981 BHs.
[ 013360.01 ] Task 43: Allocated 109.581 MB for 713031 sph, 84790 stars and 13981 BHs.
[ 013549.74 ] Task 43: Allocated 109.581 MB for 713031 sph, 84790 stars and 13981 BHs.
[ 013720.03 ] Task 43: Allocated 109.867 MB for 713031 sph, 92274 stars and 13981 BHs.
[ 013967.03 ] Task 43: Allocated 109.867 MB for 713031 sph, 92274 stars and 13981 BHs.
[ 013980.12 ] Task 43: Allocated 109.867 MB for 713031 sph, 92274 stars and 13981 BHs.
[ 014148.76 ] Task 43: Allocated 109.867 MB for 713031 sph, 92274 stars and 13981 BHs.
[ 014417.37 ] Task 43: Allocated 109.867 MB for 713031 sph, 92274 stars and 13981 BHs.
[ 014564.32 ] Task 43: Allocated 109.867 MB for 713031 sph, 92274 stars and 13981 BHs.
[ 014918.21 ] Task 43: Allocated 109.867 MB for 713031 sph, 92274 stars and 13981 BHs.
[ 014932.30 ] Task 43: Allocated 109.867 MB for 713031 sph, 92274 stars and 13981 BHs.
[ 015163.52 ] Task 43: Allocated 109.867 MB for 713031 sph, 92274 stars and 13981 BHs.
[ 015233.88 ] Task 43: Allocated 110.157 MB for 713031 sph, 99887 stars and 13981 BHs.
[ 015274.89 ] Task 43: Allocated 110.157 MB for 713031 sph, 99887 stars and 13981 BHs.
[ 015461.68 ] Task 43: Allocated 110.157 MB for 713031 sph, 99887 stars and 13981 BHs.
[ 015681.45 ] Task 43: Allocated 110.157 MB for 713031 sph, 99887 stars and 13981 BHs.
[ 015684.89 ] Task 43: Allocated 110.426 MB for 713031 sph, 106944 stars and 13981 BHs.
[ 015963.54 ] Task 43: Allocated 110.426 MB for 713031 sph, 106944 stars and 13981 BHs.
[ 015978.73 ] Task 43: Allocated 110.426 MB for 713031 sph, 106944 stars and 13981 BHs.
[ 016196.20 ] Task 43: Allocated 110.426 MB for 713031 sph, 106944 stars and 13981 BHs.
[ 016374.30 ] Task 43: Allocated 110.426 MB for 713031 sph, 106944 stars and 13981 BHs.
[ 016562.40 ] Task 43: Allocated 110.426 MB for 713031 sph, 106944 stars and 13981 BHs.
[ 016759.98 ] Task 43: Allocated 110.426 MB for 713031 sph, 106944 stars and 13981 BHs.
[ 016968.26 ] Task 43: Allocated 110.426 MB for 713031 sph, 106944 stars and 13981 BHs.

**Comment by sbird** [comment]
It is perhaps notable that 98 (!) reallocs are occurring for Task 37 (the crashing one). As far as I can see, far fewer are occurring for other tasks. We increase by 1% of MaxPart each time this happens, so this being almost the 100'th realloc may be important? If size is being increased for a type which does not have a slot somewhere, maybe something is overflowing?

**Comment by sbird** [comment]
We do set size for type 1 in slots_setup_topology so it is possible that 066eb1ddc2fc021 fixes this.

**Comment by rainwoodman** [comment]
If the type is not enabled we shouldn't have done anything to it though.

**Comment by rainwoodman** [comment]
Chris already has a stack trace but just missing line numbers. the generation of stack traces really don't belong to the application -- I feel it is marginal gain with a lot of complexity to bundle libunwind or add dependency.

Figuring out how to obtain a core dump file on the system is more useful -- e.g. we can also check the value of variables at the crash point.

**Comment by sbird** [comment]
Fair enough - I'm pretty sure there's a python script somewhere out there that can take the stack trace with addresses and the binary (ie, what we have) and turn it into line numbers. I found it once, but now seem to have misplaced it. The trouble with core dumps and MP-Gadget is that they can become really large - ideally we would dump the allocator headers and stack only.

**Comment by sbird** [comment]
Right, we do nothing with the disabled type: there should be zero bytes allocated to it. But we do do a realloc if the size is larger than maxsize, even if it is a null op because elsize=0. We are doing lots of reallocs and not changing the size of the enabled slots, so I suspect that we are increasing the size of the disabled slots.

maxsize and size are both signed, and signed overflow is undefined, so I wondered if one of them is overflowing and causing crazy things to happen (eg: maxsize overflows and the cpu writes the overflow to the low bits of ptr, causing the memmove to segfault). If this is the case, turning off -ffast-math should also cause the segfault to go away.

**Comment by sbird** [comment]
Right - I confirm that this is what is going on, at least with the reallocs: every time exchange is called info[ptype].size is increased by the number of particles exchanged, even if a slot is disabled. Eventually maxsize > size (since gc is never called for a disabled slot) and so a realloc happens. This realloc does nothing really, but I would guess that eventually maxsize overflows and undefined behaviour occurs leading to a crash.

**Comment by rainwoodman** [comment]
The problem is problematic it shouldn't have increased if slots aren't enabled for the type!

We shall probably promote size and maxsize to int64 and crash if it goes bigger than int32.

**Comment by sbird** [comment]
If an enabled slot got large we would already hit All.MaxPart (or if we have a very very threaded machine something odd would happen with NumPart). So that might be overkill.

**Comment by rainwoodman** [comment]
Sure. Never mind then.

**Comment by Chris-Pedersen** [comment]
I tried the same run with the latest commit on Simeon's slotsfix branch and get the same seg fault with:
[ 016207.25 ] tree is done.
[ 016207.28 ] Forces computed.
[ 016207.28 ] HCI: Nothing happened.
[ 016207.36 ] domain decomposition... (presently allocated=178.137 MB)
[ 016207.36 ] Allocated 10.6674 MByte for top-level domain structure
[ 016207.36 ] use of 13.3334 MB of temporary storage for domain decomposition... (presently allocated=202.156 MB)
[ 016207.41 ] local topTree size before appending=2689
[ 016207.41 ] Final local topTree size = 2825 per segment = 29.4271.
[ 016207.41 ] NTopLeaves= 2472  NTopNodes=2825 (space for 349526)
[ 016207.43 ] Expected segment cost 1.33854e+10
[ 016207.43 ] Created 96 segments in 1 rounds. Max leaf cost: 0.259138
[ 016207.43 ] Largest load: work=1.14891 particle=2.48082
[ 016207.43 ] desired memory imbalance=2.48082  (limit=699050, needed=867110)
[ 016207.43 ] Note: the domain decomposition is suboptimum because the ceiling for memory-imbalance is reached
[ 016207.43 ] Expected segment cost 349525
[ 016207.43 ] Created 96 segments in 1 rounds. Max leaf cost: 0.250002
[ 016207.43 ] Largest load: work=4.8432 particle=1.15679
[ 016207.43 ] Balance breakdown:
[ 016207.43 ] Task: [  0]  work=  1.2560  particle load=  1.0879
[ 016207.43 ] Task: [  1]  work=  0.5552  particle load=  0.8942
[ 016207.43 ] Task: [  2]  work=  0.4274  particle load=  1.0168
[ 016207.43 ] Task: [  3]  work=  0.2783  particle load=  0.9536
...
[ 016207.43 ] Task: [ 88]  work=  3.2053  particle load=  1.0591
[ 016207.43 ] Task: [ 89]  work=  0.5824  particle load=  0.8835
[ 016207.43 ] Task: [ 90]  work=  0.6438  particle load=  1.0510
[ 016207.43 ] Task: [ 91]  work=  2.2589  particle load=  0.9917
[ 016207.43 ] Task: [ 92]  work=  0.6300  particle load=  0.9387
[ 016207.43 ] Task: [ 93]  work=  0.3842  particle load=  1.0714
[ 016207.43 ] Task: [ 94]  work=  0.5318  particle load=  1.0500
[ 016207.43 ] Task: [ 95]  work=  0.7060  particle load=  0.9395
[ 016207.45 ] Using 6587309312 bytes for exchange.
[ 016207.47 ] iter=0 exchange of 0000000180042 particles
[ 016207.52 ] GC : Reducing Particle slots from 33554432 to 33374390
[ 016207.54 ] GC: Reducing number of slots for 4 from 745177 to 737030

Output files are:

[slotfix_err.txt](https://github.com/rainwoodman/MP-Gadget/files/1548554/slotfix_err.txt)
[slotfix_out.txt](https://github.com/rainwoodman/MP-Gadget/files/1548558/slotfix_out.txt)

Do you want me to speak to sysadmin about getting a core dump?

**Comment by sbird** [comment]
That looks like a different (but probably related) crash. We probably need the function names in the backtrace now. Can you send us the binary?

**Comment by rainwoodman** [comment]
What about running ```addr2line MP-Gadget 0x42c780``` and so for a few other numbers in .err file?

**Comment by sbird** [comment]
There is also something weird in that backtrace: it seems something is recursing, but there is no recursion happening in area of code. I am also experiencing odd behaviour in the exchange code, crashes with "ID mismatched" after exchange, but only with the -march=native flag.

Something weird seems to be going on here.

**Comment by Chris-Pedersen** [comment]
The job on the latest commit is at `/home/chrisp/Projects/testSims/z60_n256_10k_date12_08` on Hypatia if you can access it - otherwise I can email the binary, do you just mean the executable MP-Gadget in build?

**Comment by sbird** [comment]
Here is the function backtrace (skipping the bits in libgomp): One must use eu-addr2line.
It is crashing inside slots_gc_sorted. We had missed the important fact that this is a PM step!

[sbird@hypatia-login build]$ grep -P -o "0x4[0-9a-f]*" /home/chrisp/Projects/testSims/z60_n256_10k_date12_08/MP-Gadget.80s-105976,compute-0-21.local.btr | eu-addr2line -f -e MP-Gadget                                                                       slot_cmp_reverse_link
/home/chrisp/Codes/MP-Gadget/slotsmanager.c:218
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:131
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
msort_with_tmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:62
qsort_openmp._omp_fn.0
/home/chrisp/Codes/MP-Gadget/openmpsort.c:287
qsort_openmp
/home/chrisp/Codes/MP-Gadget/openmpsort.c:226
slots_gc_sorted
/home/chrisp/Codes/MP-Gadget/slotsmanager.c:398
domain_decompose_full
/home/chrisp/Codes/MP-Gadget/domain.c:180
run
/home/chrisp/Codes/MP-Gadget/run.c:96
main
/home/chrisp/Codes/MP-Gadget/main.c:110
_start
??:0

**Comment by sbird** [comment]
I guess that somehow the slotsmanager ptr member is being corrupted? (if it were reading past the end of the array, it would not crash, because we malloced and zerod all that memory). 
So either: a) myrealloc is returning a garbage pointer.
b) Somehow whatever comes before Base is over-writing memory in a weird way?

**Comment by sbird** [comment]
Actually, I bet at some point we are writing to SlotsManager.ptr, instead of the memory it points to. That would do it.

**Comment by rainwoodman** [comment]
I don't see anywhere we do that -- unless we triggered some compiler bugs (sounds so with -mnative?). But could be sloppy use of int and ptrdiff_t.

**Comment by rainwoodman** [comment]
```
grep '\.ptr' *.c
exchange.c:                (char*) SlotsManager->info[ptype].ptr + P[i].PI * elsize, elsize);
exchange.c:        char * ptr = SlotsManager->info[ptype].ptr;
memory.c:        memmove(newptr, tmp.ptr, tmp.size);
slotsmanager.c:        qsort_openmp(SlotsManager->info[ptype].ptr,
slotsmanager.c:            SlotsManager->info[ptype].ptr,
slotsmanager.c:        SlotsManager->info[ptype].ptr = newSlotsBase + offsets[ptype];
slotsmanager.c:    GDB_SphP = (struct sph_particle_data *) SlotsManager->info[0].ptr;
slotsmanager.c:    GDB_StarP = (struct star_particle_data *) SlotsManager->info[4].ptr;
slotsmanager.c:    GDB_BhP = (struct bh_particle_data *) SlotsManager->info[5].ptr;
```

The realloc routine in memory.c doesn't support shrinking from the top  -- do we ever do that?

**Comment by sbird** [comment]
@Chris-Pedersen : do you get the same crash if you compile with a lower optimisation level, or without -march=native?

**Comment by sbird** [comment]
@rainwoodman : We only ever grow, and only from the bottom. So it's not that.

**Comment by Chris-Pedersen** [comment]
I just noticed that in the paramfile.gadget for the last run that crashed, I have BlackHoleOn=-1 somehow - I reran with BlackHoleOn=0 and the run finished ok so it could be that which is causing it? Rerunning again to check at the moment.

**Comment by rainwoodman** [comment]
Yes BlachHoleOn=-1 will inconsistently enable blackholes here and there. So that could be the cause of crash.

**Comment by sbird** [comment]
It is also possible that this is a crash which only occurs when tree slots are allocated. Can you try again with BlackHoleOn = 1 and -DBLACK_HOLES in the Makefile, should this run complete successfully?

**Comment by Chris-Pedersen** [comment]
The run with BlackHoleOn=0 finished fine - I recompiled and ran as you said and that job just crashed with:

[ 016137.73 ] Begin force computation.
[ 016137.73 ] Start density computation...
[ 016143.89 ] ngb iteration 1: need to repeat for 137599 particles.
[ 016144.11 ] ngb iteration 2: need to repeat for 283 particles.
[ 016144.13 ] ngb iteration 3: need to repeat for 269 particles.
[ 016144.15 ] ngb iteration 4: need to repeat for 1 particles.
[ 016144.19 ] Hmax exchange: 28 toplevel tree nodes out of 2549
[ 016144.19 ] Start hydro-force computation...
[ 016154.49 ] Begin tree force.  (presently allocated=234.109 MB)
[ 016172.11 ] tree is done.
[ 016172.13 ] max number of regions is 46
[ 016172.33 ] PetaPM:  0002811871/0002985065 Pencils and 0137233138 Cells
[ 016174.68 ] Task 0: Writing Power Spectrum to output/powerspectrum-0.1584.txt
[ 016174.68 ] Tree construction.  (presently allocated=183.68 MB)
[ 016174.68 ] Allocating memory for 492247 tree-nodes (MaxPart=699050).
[ 016174.68 ] Allocated 50.4111 MByte for BH-tree, (presently allocated 234.109 MB)
[ 016174.78 ] Tree construction done.
[ 016174.78 ] Beginning black-hole accretion
[ 016174.80 ] Start swallowing of gas particles and black holes
[ 016174.82 ] Accretion done: 0 gas particles swallowed, 0 BH particles swallowed
[ 016176.71 ] SFR: spawned 0 stars, converted 1916 gas particles into stars
[ 016176.71 ] Forces computed.
[ 016176.80 ] type=0  dmean=39.0625 asmth=24.4141 minmass=7.67608e-05 a=0.158374  sqrt(<p^2>)=11.6457  dlogmax=0.1
[ 016176.81 ] type=1  dmean=39.0625 asmth=24.4141 minmass=0.000388767 a=0.158374  sqrt(<p^2>)=11.9728  dlogmax=0.00890462
[ 016176.81 ] type=4  dmean=39.0625 asmth=24.4141 minmass=7.67608e-05 a=0.158374  sqrt(<p^2>)=11.6457  dlogmax=0.00866137
[ 016176.81 ] type=5  dmean=39.0625 asmth=24.4141 minmass=7.67608e-05 a=0.158374  sqrt(<p^2>)=11.6457  dlogmax=0.00866137
[ 016176.81 ] Maximal PM timestep: dloga = 0.0051088  (0.1)
[ 016177.01 ] Attempting a domain exchange
[ 016177.03 ] Using 6581603584 bytes for exchange.
[ 016177.06 ] iter=0 exchange of 0000000002543 particles
[ 016177.11 ] GC : Reducing Particle slots from 33556348 to 33551889
[ 016177.01 ] Task 87: ngc = 0 > used = -1!


[test_out.txt](https://github.com/rainwoodman/MP-Gadget/files/1556352/test_out.txt)

**Comment by sbird** [comment]
Uh. That's a little odd. How did used get to be negative? I mean, that would explain the crash, I think...

**Comment by sbird** [comment]
Is it this line in slots_gc_sorted?
        /*Reduce slots used*/
        while(BASESLOT_PI(SlotsManager->info[ptype].size-1, ptype)->IsGarbage) {
            SlotsManager->info[ptype].size--;
        }
?
If BASESLOT_PI(0) is garbage *and* the memory before this value also happens to contain a garbage particle then we will decrement to -1 

So the fix is:
        /*Reduce slots used*/
        while(SlotsManager->info[ptype].size > 0 && BASESLOT_PI(SlotsManager->info[ptype].size-1, ptype)->IsGarbage) {
            SlotsManager->info[ptype].size--;
        }

**Comment by sbird** [comment]
And we just memmoved the data upwards, so it is more likely that the entry before will contain garbage. I think this is it.

**Comment by rainwoodman** [comment]
yes. I think so too.

**Comment by sbird** [comment]
Ok! I sent a pull request, and I am running a test myself now.

**Comment by sbird** [comment]
@Chris-Pedersen This should be fixed in the master branch

**Comment by Chris-Pedersen** [comment]
I'm still getting this error for a few runs with nGrid=128 (256 all finish OK). I tried the same sim on DiRAC too and got exactly the same crash at the same point (can upload the ouptut file from that too if necessary).

[test_out.txt](https://github.com/rainwoodman/MP-Gadget/files/1600324/test_out.txt)
[test_err.txt](https://github.com/rainwoodman/MP-Gadget/files/1600327/test_err.txt)

**Comment by sbird** [comment]
Where is the simulation output? I will see if I can take a look.

**Comment by Chris-Pedersen** [comment]
I attached the output above!

**Comment by sbird** [comment]
Aye, but I want to run eu-addr2line on the binary again :)

**Comment by rainwoodman** [comment]
What about compiling into a dynamic executable?

**Comment by rainwoodman** [comment]
Options.mk doesn't seem to indicate a static build, so I wonder why symbols were not found in the main executable. @Chris-Pedersen in parallel to that could you run ldd on MP-Gadget and see if it is dynamic?

For a dynamic executable  glibc should have pretty printed the symbol names.

**Comment by Chris-Pedersen** [comment]
Ahh woops ok! I will do tomorrow - if you want to look at the job directly before then its at:

/share/hypatia/lya_sims/initialTests/scalingTests/128_L20_lya/8_12

and there's another job with the same error at:

/share/hypatia/lya_sims/initialTests/scalingTests/128_L20_lya/4_24

**Comment by sbird** [comment]
hmm, this is a crash in domain_toptree_merge: domain_toptree_merge
/home/chrisp/Codes/MP-Gadget/domain.c:1287
domain_toptree_merge
/home/chrisp/Codes/MP-Gadget/domain.c:1332
domain_toptree_merge
/home/chrisp/Codes/MP-Gadget/domain.c:1332
domain_toptree_merge
/home/chrisp/Codes/MP-Gadget/domain.c:1332
domain_toptree_merge
/home/chrisp/Codes/MP-Gadget/domain.c:1332
domain_toptree_merge
/home/chrisp/Codes/MP-Gadget/domain.c:1332
domain_nonrecursively_combine_topTree
/home/chrisp/Codes/MP-Gadget/domain.c:1048
domain_determine_global_toptree inlined at /home/chrisp/Codes/MP-Gadget/domain.c:290 in domain_decompose_full
/home/chrisp/Codes/MP-Gadget/domain.c:1138
run
/home/chrisp/Codes/MP-Gadget/run.c:96
main
/home/chrisp/Codes/MP-Gadget/main.c:110
_start

**Comment by rainwoodman** [comment]
We are not supposed to see 1332 nested calling each other, unless TreeB is messed up such that the shift variable is bad (overflow?)

**Comment by rainwoodman** [comment]
Indeed it can be bad. I added a fix and a fancier heuristic for retrying domain decomposition. @sbird take a look it it is reasonable, if so we can deliver this for testing.

**Comment by Chris-Pedersen** [comment]
@rainwoodman when running ldd I get the output:

[chrisp@hypatia-login build]$ ldd MP-Gadget 
	linux-vdso.so.1 =>  (0x00007ffd4adab000)
	libm.so.6 => /lib64/libm.so.6 (0x0000003959800000)
	libgsl.so.23 => /share/apps/gsl-2.4/lib/libgsl.so.23 (0x00002b7a99702000)
	libgslcblas.so.0 => /share/apps/gsl-2.4/lib/libgslcblas.so.0 (0x00002b7a99b69000)
	libmpi.so.12 => /share/apps/mvapich2-2.3b/lib/libmpi.so.12 (0x00002b7a99da2000)
	libgomp.so.1 => /share/apps/gnu/gcc/5.1/lib64/libgomp.so.1 (0x00002b7a9a33b000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x0000003959400000)
	libc.so.6 => /lib64/libc.so.6 (0x0000003958c00000)
	libibverbs.so.1 => /usr/lib64/libibverbs.so.1 (0x000000395c400000)
	libpsm_infinipath.so.1 => /usr/lib64/libpsm_infinipath.so.1 (0x00002b7a9a55d000)
	libgfortran.so.3 => /share/apps/gnu/gcc/5.1/lib64/libgfortran.so.3 (0x00002b7a9a7b1000)
	libgcc_s.so.1 => /share/apps/gnu/gcc/5.1/lib64/libgcc_s.so.1 (0x00002b7a9aad2000)
	/lib64/ld-linux-x86-64.so.2 (0x0000003958800000)
	librt.so.1 => /lib64/librt.so.1 (0x0000003959c00000)
	libdl.so.2 => /lib64/libdl.so.2 (0x0000003959000000)
	libnl.so.1 => /lib64/libnl.so.1 (0x000000395ec00000)
	libinfinipath.so.4 => /usr/lib64/libinfinipath.so.4 (0x00002b7a9ace9000)
	libuuid.so.1 => /lib64/libuuid.so.1 (0x000000339e400000)
	libquadmath.so.0 => /share/apps/gnu/gcc/5.1/lib/../lib64/libquadmath.so.0 (0x00002b7a9aef9000)

**Comment by rainwoodman** [comment]
It is a dynamic executable but glibc still failed to obtain the symbol names. I confirmed this on my computer as well -- even with -rdynamic. We'll wait till #187 is merged, which adds an option -DSTACKTRACE to print a stack trace with eu-stack / pstack if exists.

**Comment by sbird** [comment]
@Chris-Pederson: try master and please add the config option -DSTACKTRACE to Options.mk

**Comment by Chris-Pedersen** [comment]
On master the sim crashes far earlier (z~25) than before (z~2.8) with:

[ 000131.96 ] Forces computed.
[ 000131.96 ] Task 88: [ 000131.96 ] bad timestep spotted: terminating and saving snapshot.
Bad timestep (1) assigned! ID=3136082 Type=0 dloga=3.04045e-06 dtmax=8000 xyz=(9703.1|8095.58|12405.6) tree=(-0.0216029|-0.0156799|-0.0129779) PM=(-0.344814|0.308484|-0.837693)
[ 000131.96 ] Task 88: hydro-frc=(196440|2.27784e+06|-237381) dens=2.18748e-09 hsml=260.413 numngb=32.9384 egyrho=2.18748e-09 dhsmlegydensityfactor=1.03365 Entropy=2.36864e+25, dtEntropy=4.41443e+28
[ 000131.96 ] Task 88: Time-step of integer size 1 not allowed, id = 3136082, debugging info follows. 2

[test_err.txt](https://github.com/rainwoodman/MP-Gadget/files/1618805/test_err.txt)
[test_out.txt](https://github.com/rainwoodman/MP-Gadget/files/1618806/test_out.txt)

Other sims that finished OK before are also crashing early with the same error.

**Comment by sbird** [comment]
Yes, back it out until I have time to fix it.

**Comment by Chris-Pedersen** [comment]
Appears to be fixed, sims that were seg faulting before are finishing fine now.
